File: Data Structures & Algorithms/two-integer-sum/submission-0.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        dictionary = dict()
        for i in range(0, len(nums)):
            want = target - nums[i]
            if dictionary.get(want) != None:
                return [dictionary.get(want), i]
            dictionary[nums[i]] = i

        return [-1, -1]

File: Data Structures & Algorithms/combination-target-sum/submission-0.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        set<vector<int>> ans;
        solve(0, nums, target, {}, ans);
        return vector<vector<int>>(ans.begin(), ans.end());
    }

    void solve(int index, vector<int> &nums, int target, vector<int> curr, set<vector<int>> &ans)
    {
        if (target == 0)
        {
            ans.insert(curr);
        }

        if (index == nums.size() || target < 0) return ;

        curr.push_back(nums[index]);
        solve(index, nums, target - nums[index], curr, ans);
        curr.pop_back();
        solve(index + 1, nums, target, curr, ans);
    }
};
